File: Week 1/rust/9.data_types.rs
/*
 * WEEK 1 - RUST FUNDAMENTALS
 * Topic: Data Types in Rust
 * File: 9.data_types.rs
 *
 * CONCEPT:
 * Rust's scalar (primitive) types are explicit and fixed-width:
 *   Signed integers:  i8, i16, i32, i64, i128, isize  (isize == ptr width)
 *   Unsigned ints:    u8, u16, u32, u64, u128, usize
 *   Floats:           f32, f64
 *   Boolean:          bool       (true / false)
 *   Character:        char       (4 bytes, Unicode scalar value)
 *   Unit:             ()         (used like Java's `void`)
 *
 * Compound types: tuple `(T1, T2)`, array `[T; N]`, slice `&[T]`, str / String.
 *
 * KEY POINTS:
 *  - Default integer type: i32; default float: f64.
 *  - `usize` / `isize` are used for indexing (platform-dependent width).
 *  - `char` is 4 bytes (any Unicode scalar value), NOT 1 byte ASCII.
 *  - `String` is a growable, heap-allocated, UTF-8 sequence.
 *  - `&str` is a borrowed string slice (fixed-size view into a UTF-8 buffer).
 *  - `bool` is exactly 1 byte but logically `true`/`false`.
 *  - `std::mem::size_of::<T>()` returns the byte size.
 *
 * SYNTAX:
 *   let i: i32 = 42;
 *   let f: f64 = 3.14;
 *   let b: bool = true;
 *   let c: char = 'A';            // 'A' as char (NOT u8)
 *   let s: &str = "hello";        // string slice (Display-friendly)
 *   let owned: String = String::from("world");
 *
 * DRY RUN:
 *  size_of::<i32>() = 4
 *  size_of::<f64>() = 8
 *  size_of::<char>() = 4
 *  size_of::<bool>() = 1
 */

fn main() {
    use std::mem::size_of;

    println!(
        "{:<12}{:<8}",
        "type", "bytes"
    );
    println!("--------------------");
    println!("{:<12}{:<8}", "i8",    size_of::<i8>());
    println!("{:<12}{:<8}", "i16",   size_of::<i16>());
    println!("{:<12}{:<8}", "i32",   size_of::<i32>());
    println!("{:<12}{:<8}", "i64",   size_of::<i64>());
    println!("{:<12}{:<8}", "i128",  size_of::<i128>());
    println!("{:<12}{:<8}", "isize", size_of::<isize>());
    println!("{:<12}{:<8}", "u32",   size_of::<u32>());
    println!("{:<12}{:<8}", "f32",   size_of::<f32>());
    println!("{:<12}{:<8}", "f64",   size_of::<f64>());
    println!("{:<12}{:<8}", "bool",  size_of::<bool>());
    println!("{:<12}{:<8}", "char",  size_of::<char>());
    println!("{:<12}{:<8}", "()",    size_of::<()>());

    let max_i32 = i32::MAX;
    let min_i32 = i32::MIN;
    println!("\ni32::MAX = {max_i32}, i32::MIN = {min_i32}");

    // No overflow for i128 unless you EXCEED its range; otherwise it panics in debug.
    let big: i128 = i128::MAX;
    println!("i128::MAX = {big}");

    // char vs u8
    let ch: char = 'Z';
    println!("char 'Z' code point = {}", ch as u32);

    // Tuple and array
    let tup: (i32, f64, char) = (1, 2.0, '3');
    let arr: [i32; 4] = [10, 20, 30, 40];
    println!("tuple = {tup:?}, array = {arr:?}");
}

/*
 * NOTES:
 *  - Java has 32-bit int and 64-bit long; Rust spells these i32 / i64.
 *  - Java's char is UTF-16 (2 bytes); Rust's char is full Unicode (4 bytes).
 *  - Rust has NO implicit conversions between numeric types — use `as`.
 *  - Use `{:?}` (Debug) to print compound types like tuples and arrays.
 *  - `String` is owning + growable; `&str` is borrowed + read-only.
 */

File: Week 1/python/9.data_types.py
"""
WEEK 1 - PYTHON FUNDAMENTALS
Topic: Data Types in Python
File: 9.data_types.py

CONCEPT:
Java has 8 PRIMITIVE types (byte, short, int, long, float, double, char, boolean)
plus reference types like String. Python has NO primitives — everything is an
object. Its main built-in types are:
  int, float, complex, bool, str, bytes, list, tuple, dict, set, NoneType

KEY POINTS:
- `int`   : arbitrary precision (no overflow). Replaces byte/short/int/long.
- `float` : 64-bit IEEE 754. Replaces both Java float AND double.
- `complex`: a + bj  (no Java equivalent in the standard library)
- `bool`  : True / False; subclass of int — True == 1, False == 0
- `str`   : immutable Unicode string. Replaces both Java String AND char.
- `bytes` : immutable sequence of 0..255 ints (use for binary data)
- `NoneType`: the singleton `None`, similar to Java's null
- `type(x)` returns x's class. `isinstance(x, T)` checks type.

PYTHON-SPECIFIC NOTES vs Java:
- No `char` type — single characters are length-1 strings.
- `str` is immutable, just like Java's String.
- Use `sys.getsizeof(obj)` to inspect runtime size (in bytes).

SYNTAX:
n: int   = 42
x: float = 3.14
b: bool  = True
s: str   = "hello"
none     = None

DRY RUN:
1. type(42)        -> <class 'int'>
2. type(3.14)      -> <class 'float'>
3. type("hi")      -> <class 'str'>
4. type(True)      -> <class 'bool'>
5. 2 ** 100        -> 1267650600228229401496703205376  (no overflow!)
"""

import sys


def show_types() -> None:
    samples = [
        42,
        3.14,
        2 + 3j,
        True,
        False,
        "hello",
        b"bytes",
        [1, 2, 3],
        (1, 2, 3),
        {"a": 1},
        {1, 2, 3},
        None,
    ]
    print(f"{'value':<20} {'type':<20} {'size (bytes)':<15}")
    print("-" * 55)
    for v in samples:
        print(f"{repr(v):<20} {type(v).__name__:<20} {sys.getsizeof(v):<15}")


def show_no_overflow() -> None:
    """Python ints have arbitrary precision — Java int would overflow at 2^31."""
    print("\nLarge integer arithmetic (no overflow!):")
    big = 2 ** 100
    print(f"2^100  = {big}")
    print(f"100!   = {factorial(100)}")


def factorial(n: int) -> int:
    """Iterative factorial; returns Python int (arbitrary precision)."""
    result = 1
    for i in range(2, n + 1):
        result *= i
    return result


def show_string_is_char() -> None:
    """Demonstrate that there is no separate `char` type."""
    s = "Z"
    print(f"\nSingle character behaviour:")
    print(f"  s        = {s!r}")
    print(f"  type(s)  = {type(s).__name__}")
    print(f"  len(s)   = {len(s)}")
    print(f"  ord(s)   = {ord(s)}    # Java: (int) 'Z' -> 90")
    print(f"  chr(90)  = {chr(90)!r}  # Java: (char) 90 -> 'Z'")


def main() -> None:
    show_types()
    show_no_overflow()
    show_string_is_char()


if __name__ == "__main__":
    main()


# NOTES:
# - Java ints/floats are PRIMITIVES living on the stack; Python wraps them as
#   objects on the heap (CPython interns small ints for speed).
# - Java's `String` is reference type but immutable; Python's `str` is the same.
# - Java has separate `char` (16-bit Unicode); Python uses 1-character strings.
# - Booleans in Python are ints — `True + True == 2`. Java forbids this.
# - There is no `null` keyword — Python uses `None`. Comparison: `x is None`.
